Improvements to image stacking

graeme-winter opened the pull request:
Means the images are on a sensible scale when stacked so look OK

graeme-winter commented:
Before:

![Screenshot 2020-06-12 at 16 05 13](https://user-images.githubusercontent.com/1241716/84517063-a2059b80-acc6-11ea-9fea-74b3047304ba.png)

adding 10 images => 

![Screenshot 2020-06-12 at 16 05 00](https://user-images.githubusercontent.com/1241716/84517083-a8941300-acc6-11ea-93e7-7fafff74caed.png)


After:

adding 10 images gives:

![Screenshot 2020-06-12 at 16 05 46](https://user-images.githubusercontent.com/1241716/84517113-b184e480-acc6-11ea-93bf-11d7a5b8aac8.png)


which I think is much more what you would want?

dagewa commented:
The function name `sum_images` is not really appropriate if the result is actually to calculate a mean

graeme-winter commented:
@dagewa valid - but don't want to rename it to `compute_mean_over_N_images` or something? 🤔

graeme-winter commented:
Set max brightness to 1000 to reproduce feel of stacking 10 images as before

graeme-winter commented:
While I am thinking about this, wonder if we should have a show-max mode as well? 🤔  so maximum pixel over the range rather than mean value?

dagewa commented:
Could definitely be useful - has been used for virtual powder patterns I think

rjgildea commented on util/image_viewer/spotfinder_frame.py:
```diff
@@ -811,7 +811,7 @@ def draw_resolution_rings(self, unit_cell=None, space_group=None):
 
-    def sum_images(self):
+    def stack_images(self):
         if self.params.sum_images > 1:
```
Would it make sense to also change the command line parameter to `stack_images` (possibly keeping around `sum_images` with a deprecation warning for the time being)?

graeme-winter replied:
... maybe, provided the people's national bikeshed committee are satisfied with the nomenclature of stacking the images?

graeme-winter replied:
I don't know how to make one phil parameter alias another, and deprecate one or the other, otherwise I would have done this...

graeme-winter replied:
I could replace the name and just break the cl api

rjgildea replied:
R.e. marking phil parameters as deprecated:
https://github.com/cctbx/cctbx_project/issues/452#issuecomment-604313748

I would either:
1. Rename `sum_images` to `stack_images` and re-add `sum_images` with `.deprecated` attribute, and ignore. That way it doesn't fail, but users get a warning.
2. Just rename and accept that very few people likely use the CLI to the image viewer.

graeme-winter replied:
Pushed change to this effect, if you use old command line option you get 

```
Grey-Area 0 :) $ dials.image_viewer imported.expt strong.refl sum_images=10
DIALS (2018) Acta Cryst. D74, 85-97. https://doi.org/10.1107/S2059798317017235
Sorry: Unknown command line parameter definition: sum_images = 10
```

which I think is OK

ndevenish commented:
I worry that if summing over a range to try to find weak signal, or signal only present on a few, that this might hide it? A "max pixel" mode might be appropriate. Maybe this does deserve a setting, but maybe that's just avoiding making a decision...

graeme-winter commented:
> I worry that if summing over a range to try to find weak signal, or signal only present on a few, that this might hide it? A "max pixel" mode might be appropriate. Maybe this does deserve a setting, but maybe that's just avoiding making a decision...

My thinking was about simulating a wider image in the first instance, but the max is useful as you say for getting the pattern out. I think the universe has space for both...

graeme-winter commented:
@ndevenish added a mean / max pulldown - illustrated difference:

Max:

![Screenshot 2020-06-15 at 09 12 18](https://user-images.githubusercontent.com/1241716/84633533-76fa9200-aee8-11ea-977f-1eac2fab4726.png)


Mean:

![Screenshot 2020-06-15 at 09 12 41](https://user-images.githubusercontent.com/1241716/84633542-7bbf4600-aee8-11ea-8652-b0d06acdfa03.png)

rjgildea commented:
As far as I can tell, you can only change the number of images to stack via the command line but not via the GUI, and conversely, you can only change the stack "type" in the GUI but not via the command line. This means that if you load the image viewer with default options, then there is a "stack type" dropdown that doesn't do anything.

The `stack_type` setting should probably be exposed via the command line, and it only makes really makes sense to have the "stack type" dropdown in the GUI settings panel if you can also control the number of images there too.

rjgildea commented:
This seems incompatible with setting the "image type" to be "raw":

```
$ dials.image_viewer $(dials.data get -q x4wide)/*.cbf stack_images=10
```
Select "raw" in the "image type" dropdown:
```
Traceback (most recent call last):
  File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 2140, in OnUpdateImage
    self.GetParent().GetParent().reload_image()
  File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 513, in reload_image
    self.load_image(self.images.selected, refresh=True)
  File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 558, in load_image
    show_untrusted=show_untrusted,
  File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/slip_viewer/frame.py", line 499, in load_image
    self.update_settings()
  File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 1175, in update_settings
    self.stack_images()
  File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 836, in stack_images
    sel = data > rd
TypeError: '>' not supported between instances of 'double' and 'int'
```

rjgildea commented:
Another oddity, this behaviour was probably at least in part already present before this PR - if you're in stack/sum images mode, then press the "image" button, then the displayed image goes back to the individual, rather than stacked, image, until you re-select an item from the "stack type" dropdown menu. I'm not entirely sure what the expected should be for the various thresholding maps - should we be running the threshold algorithms on the stacked image?

graeme-winter commented:
> As far as I can tell, you can only change the number of images to stack via the command line but not via the GUI, and conversely, you can only change the stack "type" in the GUI but not via the command line. This means that if you load the image viewer with default options, then there is a "stack type" dropdown that doesn't do anything.
> 
> The `stack_type` setting should probably be exposed via the command line, and it only makes really makes sense to have the "stack type" dropdown in the GUI settings panel if you can also control the number of images there too.

There is a "Stack" box at the top of the image window? I've been using this extensively so know that works 😉

I had wondered whether the `stack_type` should be exposed via command line, but not followed up on.

graeme-winter commented:
> This seems incompatible with setting the "image type" to be "raw":
> 
> ```
> $ dials.image_viewer $(dials.data get -q x4wide)/*.cbf stack_images=10
> ```
> 
> Select "raw" in the "image type" dropdown:
> 
> ```
> Traceback (most recent call last):
>   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 2140, in OnUpdateImage
>     self.GetParent().GetParent().reload_image()
>   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 513, in reload_image
>     self.load_image(self.images.selected, refresh=True)
>   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 558, in load_image
>     show_untrusted=show_untrusted,
>   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/slip_viewer/frame.py", line 499, in load_image
>     self.update_settings()
>   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 1175, in update_settings
>     self.stack_images()
>   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 836, in stack_images
>     sel = data > rd
> TypeError: '>' not supported between instances of 'double' and 'int'
> ```

Reproduced this, will look

graeme-winter commented:
> Another oddity, this behaviour was probably at least in part already present before this PR - if you're in stack/sum images mode, then press the "image" button, then the displayed image goes back to the individual, rather than stacked, image, until you re-select an item from the "stack type" dropdown menu. I'm not entirely sure what the expected should be for the various thresholding maps - should we be running the threshold algorithms on the stacked image?

Thresholding should be fine against summed, mean requires some discussion re: "gain" equivalent, max just plain wrong. Not sure if this should be imposed. 

Re: the oddity - the internals of the image viewer are something of a tangled mess. Should fixing this be a necessary precondition of the PR features being made available?

graeme-winter commented:
> > This seems incompatible with setting the "image type" to be "raw":
> > ```
> > $ dials.image_viewer $(dials.data get -q x4wide)/*.cbf stack_images=10
> > ```
> > 
> > 
> > Select "raw" in the "image type" dropdown:
> > ```
> > Traceback (most recent call last):
> >   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 2140, in OnUpdateImage
> >     self.GetParent().GetParent().reload_image()
> >   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 513, in reload_image
> >     self.load_image(self.images.selected, refresh=True)
> >   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 558, in load_image
> >     show_untrusted=show_untrusted,
> >   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/slip_viewer/frame.py", line 499, in load_image
> >     self.update_settings()
> >   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 1175, in update_settings
> >     self.stack_images()
> >   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 836, in stack_images
> >     sel = data > rd
> > TypeError: '>' not supported between instances of 'double' and 'int'
> > ```
> 
> Reproduced this, will look

Fixed / worked around in c3d54cc1f86fb2f65e03198e157b1dd48908ad86

graeme-winter commented:
> > As far as I can tell, you can only change the number of images to stack via the command line but not via the GUI, and conversely, you can only change the stack "type" in the GUI but not via the command line. This means that if you load the image viewer with default options, then there is a "stack type" dropdown that doesn't do anything.
> > The `stack_type` setting should probably be exposed via the command line, and it only makes really makes sense to have the "stack type" dropdown in the GUI settings panel if you can also control the number of images there too.
> 
> There is a "Stack" box at the top of the image window? I've been using this extensively so know that works 😉
> 
> I had wondered whether the `stack_type` should be exposed via command line, but not followed up on.

Done, in a2910e0041898ac0a3dabdf407bd4ad6a30a2950

graeme-winter commented:
> > Another oddity, this behaviour was probably at least in part already present before this PR - if you're in stack/sum images mode, then press the "image" button, then the displayed image goes back to the individual, rather than stacked, image, until you re-select an item from the "stack type" dropdown menu. I'm not entirely sure what the expected should be for the various thresholding maps - should we be running the threshold algorithms on the stacked image?
> 
> Thresholding should be fine against summed, mean requires some discussion re: "gain" equivalent, max just plain wrong. Not sure if this should be imposed.
> 
> Re: the oddity - the internals of the image viewer are something of a tangled mess. Should fixing this be a necessary precondition of the PR features being made available?

I have looked at the code for a while and can't see where the change should go to make this oddity go away - however since it is not a _new_ oddity I am not too worried about this. I have made sum the default, so if someone stacks images and then looks through the filters the results should be by default meaningful.

rjgildea commented:
> There is a "Stack" box at the top of the image window? I've been using this extensively so know that works 😉

Ah ok - I had missed that because I was naively expecting related controls to be in the _same_ window at the very least. Does it really make sense for the control for number of images to stack and the control for stack type to be in different windows?

graeme-winter commented:
> > There is a "Stack" box at the top of the image window? I've been using this extensively so know that works 😉
> 
> Ah ok - I had missed that because I was naively expecting related controls to be in the _same_ window at the very least. Does it really make sense for the control for number of images to stack and the control for stack type to be in different windows?

Respectfully acknowledged. However also mention that stacking images is related to loading images, which is all controlled above the image... and selecting the view kinda lives with the other settings, so maybe (just maybe) the viewer is poorly designed 🤷‍♂️

rjgildea commented on util/image_viewer/spotfinder_frame.py:
```diff
@@ -823,12 +824,24 @@ def sum_images(self):
-                    rd += image_data_i[j]
+                    data = image_data_i[j]
+                    if mode == "max":
+                        if type(rd) != type(data):
```
I think you can call `.as_double()` even on an array that is already `flex.double()`, so you could avoid having to explicitly check the type and always both sides to double.

`a = flex.random_double(10).as_double()`

graeme-winter commented:
> > I have looked at the code for a while and can't see where the change should go to make this oddity go away - however since it is not a _new_ oddity I am not too worried about this. I have made sum the default, so if someone stacks images and then looks through the filters the results should be by default meaningful.
> 
> I'm not convinced the thresholding maps are acting on the sum/mean/max images in any case - I suspect they are effectively switching to `stack_images=1` without updating any of the controls to reflect this.

Quite possibly: perhaps this is also in the realm of 🤷‍♂️  - since stack images existed before this I am not sure it's a new bug...

rjgildea commented:
> > > This seems incompatible with setting the "image type" to be "raw":
> > > ```
> > > $ dials.image_viewer $(dials.data get -q x4wide)/*.cbf stack_images=10
> > > ```
> > > 
> > > 
> > > Select "raw" in the "image type" dropdown:
> > > ```
> > > Traceback (most recent call last):
> > >   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 2140, in OnUpdateImage
> > >     self.GetParent().GetParent().reload_image()
> > >   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 513, in reload_image
> > >     self.load_image(self.images.selected, refresh=True)
> > >   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 558, in load_image
> > >     show_untrusted=show_untrusted,
> > >   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/slip_viewer/frame.py", line 499, in load_image
> > >     self.update_settings()
> > >   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 1175, in update_settings
> > >     self.stack_images()
> > >   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 836, in stack_images
> > >     sel = data > rd
> > > TypeError: '>' not supported between instances of 'double' and 'int'
> > > ```
> > 
> > 
> > Reproduced this, will look
> 
> Fixed / worked around in [c3d54cc](https://github.com/dials/dials/commit/c3d54cc1f86fb2f65e03198e157b1dd48908ad86)

This still breaks for me:
```
$ dials.image_viewer $(dials.data get -q x4wide)/*.cbf stack_images=10
DIALS (2018) Acta Cryst. D74, 85-97. https://doi.org/10.1107/S2059798317017235
The following parameters have been modified:

stack_images = 10
input {
  experiments = <image files>
}

Traceback (most recent call last):
  File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 2141, in OnUpdateImage
    self.GetParent().GetParent().reload_image()
  File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 513, in reload_image
    self.load_image(self.images.selected, refresh=True)
  File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 558, in load_image
    show_untrusted=show_untrusted,
  File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/slip_viewer/frame.py", line 499, in load_image
    self.update_settings()
  File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 1177, in update_settings
    self.stack_images()
  File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 839, in stack_images
    rd += data
Boost.Python.ArgumentError: Python argument types in
    int.__iadd__(int, double)
did not match C++ signature:
    __iadd__(scitbx::af::versa<int, scitbx::af::flex_grid<scitbx::af::small<long, 10ul> > > {lvalue}, int)
    __iadd__(scitbx::af::versa<int, scitbx::af::flex_grid<scitbx::af::small<long, 10ul> > > {lvalue}, scitbx::af::versa<int, scitbx::af::flex_grid<scitbx::af::small<long, 10ul> > >)```

rjgildea commented:
Also "max" stack mode + "raw" image type looks very much like it is displaying the single image, and not a stacked image.

graeme-winter commented:
<- is wondering about abandoning this project as to fix all the problems here will probably require rewriting the image viewer.

graeme-winter commented:
> Also "max" stack mode + "raw" image type looks very much like it is displaying the single image, and not a stacked image.

Sometimes the raw image is ints, sometimes doubles, sometimes the image you are adding is ints, sometimes doubles, and they need to be added in-place so will need to cast images-for-adding to the same type as image-already-there - cannot stick with one "standard" type - so this will be some horrible hodge-podge 😞

graeme-winter commented:
> > > > This seems incompatible with setting the "image type" to be "raw":
> > > > ```
> > > > $ dials.image_viewer $(dials.data get -q x4wide)/*.cbf stack_images=10
> > > > ```
> > > > 
> > > > 
> > > > Select "raw" in the "image type" dropdown:
> > > > ```
> > > > Traceback (most recent call last):
> > > >   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 2140, in OnUpdateImage
> > > >     self.GetParent().GetParent().reload_image()
> > > >   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 513, in reload_image
> > > >     self.load_image(self.images.selected, refresh=True)
> > > >   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 558, in load_image
> > > >     show_untrusted=show_untrusted,
> > > >   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/slip_viewer/frame.py", line 499, in load_image
> > > >     self.update_settings()
> > > >   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 1175, in update_settings
> > > >     self.stack_images()
> > > >   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 836, in stack_images
> > > >     sel = data > rd
> > > > TypeError: '>' not supported between instances of 'double' and 'int'
> > > > ```
> > > 
> > > 
> > > Reproduced this, will look
> > 
> > 
> > Fixed / worked around in [c3d54cc](https://github.com/dials/dials/commit/c3d54cc1f86fb2f65e03198e157b1dd48908ad86)
> 
> This still breaks for me:
> 
> ```
> $ dials.image_viewer $(dials.data get -q x4wide)/*.cbf stack_images=10
> DIALS (2018) Acta Cryst. D74, 85-97. https://doi.org/10.1107/S2059798317017235
> The following parameters have been modified:
> 
> stack_images = 10
> input {
>   experiments = <image files>
> }
> 
> Traceback (most recent call last):
>   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 2141, in OnUpdateImage
>     self.GetParent().GetParent().reload_image()
>   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 513, in reload_image
>     self.load_image(self.images.selected, refresh=True)
>   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 558, in load_image
>     show_untrusted=show_untrusted,
>   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/slip_viewer/frame.py", line 499, in load_image
>     self.update_settings()
>   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 1177, in update_settings
>     self.stack_images()
>   File "/Users/rjgildea/software/cctbx/modules/dials/util/image_viewer/spotfinder_frame.py", line 839, in stack_images
>     rd += data
> Boost.Python.ArgumentError: Python argument types in
>     int.__iadd__(int, double)
> did not match C++ signature:
>     __iadd__(scitbx::af::versa<int, scitbx::af::flex_grid<scitbx::af::small<long, 10ul> > > {lvalue}, int)
>     __iadd__(scitbx::af::versa<int, scitbx::af::flex_grid<scitbx::af::small<long, 10ul> > > {lvalue}, scitbx::af::versa<int, scitbx::af::flex_grid<scitbx::af::small<long, 10ul> > >)```
> ```

... try with 372120fc16eb07ee3dc3eb251ae9e9e9ff69dbc1

graeme-winter commented:
> > I have looked at the code for a while and can't see where the change should go to make this oddity go away - however since it is not a _new_ oddity I am not too worried about this. I have made sum the default, so if someone stacks images and then looks through the filters the results should be by default meaningful.
> 
> I'm not convinced the thresholding maps are acting on the sum/mean/max images in any case - I suspect they are effectively switching to `stack_images=1` without updating any of the controls to reflect this.

I'm pretty sure this is properly broken. Could disable these buttons if `stack != 1`? Alternative will be rewriting the way that it works.

graeme-winter commented:
@rjgildea disabled the buttons if stack now - so behaviour is consistent at least!

graeme-winter commented:
Summary:
 - using double for all image data (raw; corrected) so internally consistent
 - if stack: disable image filter buttons
 - command-line control of stack number, stack mode
 - tests with raw and corrected images work
